### PR TITLE
Modernizing the codebase - and, or, not

### PR DESCRIPTION
**Pull request by JanVogelsang**
In this PR all occurrences of `&&`, `||` and `!` are changed to `and`, `or` and `not` respectively.

Additionally, all redundant parentheses around boolean expressions are removed.

e.g.: `if ( ( a == 1 ) && ( b == 2 ) )` was changed to `if ( a == 1 and b == 2 )`

### CODE REVIEW

**Comment by JanVogelsang** [comment]
Unfortunately, two of the checks do not pass yet, because of some SLI regression test failing. I went over every single of my changes in this PR multiple times now and could not find any change that would actually change the compiled code in any way. All changes are purely style-based. Either I am missing something, or there is some reason related to SLI that I am not aware of, that causes the checks to fail. Could someone assist me here?

**Comment by heplesser** [comment]
@JanVogelsang I can at least explain why only two of the tests fail: The passing ones include MPI, and the test is skipped for MPI (for technical reasons). Can you reproduce the test failure on your computer if you build without MPI?

File: libnestutil/lockptr.h [hunk 1]
```diff
@@ -114,7 +114,7 @@ class lockPTR
     ~PointerObject()
     {
       assert( not locked );
-      if ( ( pointee != NULL ) && deletable && ( not locked ) )
+      if ( ( pointee != NULL ) and deletable and not locked )
```
Comment (heplesser) [inline libnestutil/lockptr.h:117]: Why is there still a `NULL` here?

**Comment by JanVogelsang** [comment]
I can reproduce the failing tests on my computer without building with MPI. Same error

File: sli/tokenstack.h [hunk 2]
```diff
@@ -160,7 +160,7 @@ class TokenStack : private TokenArrayObj
   void
   roll( size_t n, long k )
   {
-    if ( n < 2 || k == 0 )
+    if ( n < 2 and k == 0 )
```
Comment (heplesser) [inline sli/tokenstack.h:163]:
```suggestion
    if ( n < 2 or k == 0 )
```

**Review by heplesser — CHANGES_REQUESTED** [review]
I think I found the error, see suggestion.

File: sli/tokenstack.h (hunk 2 above)
Comment (JanVogelsang) [reply sli/tokenstack.h]: Awesome, that's it! No idea how that happened, but glad you found it. Checks pass now locally already.

File: libnestutil/lockptr.h (hunk 1 above)
Comment (JanVogelsang) [reply libnestutil/lockptr.h:117]: Fixed all of these cases in a separate PR, see #2499

Comment (heplesser) [reply libnestutil/lockptr.h:117]: OK.

**Review by heplesser — APPROVED** [review]

File: models/ac_generator.cpp [hunk 3]
```diff
@@ -202,7 +202,7 @@ nest::ac_generator::pre_run_hook()
 void
 nest::ac_generator::update( Time const& origin, const long from, const long to )
 {
-  assert( to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( to >= 0 and ( delay ) from < kernel().connection_manager.get_min_delay() );
```
Comment (jougs) [inline models/ac_generator.cpp:205]: Do plan to replace C-style casts like the one here (`( delay ) from`) to C++ `static_cast`s? If not, can you do this here?

File: libnestutil/lockptr.h (hunk 1 above)
Comment (jougs) [inline libnestutil/lockptr.h:117]: Why is the `NULL` still here?

File: models/aeif_psc_alpha.cpp [hunk 4]
```diff
@@ -438,7 +438,7 @@ nest::aeif_psc_alpha::pre_run_hook()
 void
 nest::aeif_psc_alpha::update( Time const& origin, const long from, const long to )
 {
-  assert( to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( to >= 0 and ( delay ) from < kernel().connection_manager.get_min_delay() );
```
Comment (jougs) [inline models/aeif_psc_alpha.cpp:441]: C-style cast, see above.

File: models/noise_generator.cpp [hunk 5]
```diff
@@ -289,7 +289,7 @@ nest::noise_generator::send_test_event( Node& target, rport receptor_type, synin
 void
 nest::noise_generator::update( Time const& origin, const long from, const long to )
 {
-  assert( to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( to >= 0 and ( delay ) from < kernel().connection_manager.get_min_delay() );
```
Comment (jougs) [inline models/noise_generator.cpp:292]: C-style cast, see above.

File: sli/tarrayobj.cc [hunk 6]
```diff
@@ -630,7 +630,7 @@ TokenArrayObj::operator==( const TokenArrayObj& a ) const
   // experimentally replaced by line below 090120, Diesmann
   // because [] cvx has non NULL p
   //
-  //    if( p == NULL || a.p == NULL || size() != a.size())
+  //    if( p == NULL or a.p == NULL or size() != a.size())
```
Comment (jougs) [inline sli/tarrayobj.cc:633]:
```suggestion
  //    if( not p or not a.p or size() != a.size())
```

But probably this comment can go away completely...

File: models/hh_cond_beta_gap_traub.cpp [hunk 7]
```diff
@@ -552,8 +552,8 @@ nest::hh_cond_beta_gap_traub::update_( Time const& origin,
         --S_.r_;
       }
       else
-        // (    threshold    &&     maximum       )
-        if ( S_.y_[ State_::V_M ] >= P_.V_T + 30. && U_old > S_.y_[ State_::V_M ] )
+        // (    threshold    and     maximum       )
+        if ( S_.y_[ State_::V_M ] >= P_.V_T + 30. and U_old > S_.y_[ State_::V_M ] )
```
Comment (jougs) [inline models/hh_cond_beta_gap_traub.cpp:556]: I think this should be collapsed into an `else if` on a single line...

File: models/hh_psc_alpha.cpp [hunk 8]
```diff
@@ -437,8 +437,8 @@ nest::hh_psc_alpha::update( Time const& origin, const long from, const long to )
       --S_.r_;
     }
     else
-      // (    threshold    &&     maximum       )
-      if ( S_.y_[ State_::V_M ] >= 0 && U_old > S_.y_[ State_::V_M ] )
+      // (    threshold    and     maximum       )
+      if ( S_.y_[ State_::V_M ] >= 0 and U_old > S_.y_[ State_::V_M ] )
```
Comment (jougs) [inline models/hh_psc_alpha.cpp:441]: `else if` on single line, see above.

File: nestkernel/mask_impl.h [hunk 9]
```diff
@@ -105,7 +105,7 @@ BoxMask< D >::outside( const Box< D >& b ) const
   // so we don't know if it is an actual problem.
   for ( int i = 0; i < D; ++i )
   {
-    if ( ( b.upper_right[ i ] < min_values_[ i ] ) || ( b.lower_left[ i ] > max_values_[ i ] ) )
+    if ( ( b.upper_right[ i ] < min_values_[ i ] ) or ( b.lower_left[ i ] > max_values_[ i ] ) )
```
Comment (jougs) [inline nestkernel/mask_impl.h:108]: Wouldn't this version with less parentheses be equivalent? The same comments applies to other conditions within this file. 

```suggestion
    if ( b.upper_right[ i ] < min_values_[ i ] or  b.lower_left[ i ] > max_values_[ i ] )
```

File: sli/sliarray.cc [hunk 10]
```diff
@@ -907,7 +907,7 @@ SLIArrayModule::ArraycreateFunction::execute( SLIInterpreter* i ) const
   const Token mark_token( new LiteralDatum( i->mark_name ) );
   bool found = false;
 
-  while ( ( n < depth ) && not found )
+  while ( ( n < depth ) and not found )
```
Comment (jougs) [inline sli/sliarray.cc:910]:
```suggestion
  while ( n < depth and not found )
```

File: sli/sliarray.cc [hunk 11]
```diff
@@ -2739,7 +2739,7 @@ SLIArrayModule::GaborFunction::execute( SLIInterpreter* i ) const
     i->raiseerror( "RangeCheck" );
     return;
   }
-  if ( ( ncol < 2 ) || ( nrow < 2 ) )
+  if ( ( ncol < 2 ) or ( nrow < 2 ) )
```
Comment (jougs) [inline sli/sliarray.cc:2742]:
```suggestion
  if ( ncol < 2 or nrow < 2 )
```

File: sli/sliarray.cc [hunk 12]
```diff
@@ -2847,7 +2847,7 @@ SLIArrayModule::Gauss2dFunction::execute( SLIInterpreter* i ) const
...
```
Comment (jougs) [inline sli/sliarray.cc:2850]:
```suggestion
  if ( ncol < 2 or nrow < 2 )
```

File: sli/slicontrol.cc [hunk 13]
```diff
@@ -1429,16 +1429,16 @@ SwitchdefaultFunction::execute( SLIInterpreter* i ) const
     throw TypeMismatch( "At least 1 argument.", "Nothing." );
   }
 
-  if ( depth > 1 && i->OStack.pick( 1 ) != mark_token // default action
-    && i->OStack.pick( 0 ) != mark_token )            // is not the only one
+  if ( depth > 1 and i->OStack.pick( 1 ) != mark_token // default action
+    and i->OStack.pick( 0 ) != mark_token )            // is not the only one
```
Comment (jougs) [inline sli/slicontrol.cc:1433]: Could probably go on a single line with <120 characters

File: sli/sligraphics.cc [hunk 14]
```diff
@@ -193,13 +193,13 @@ SLIgraphics::ReadPGMFunction::readImage( std::istream* in,
     if ( std::string( magic ) == std::string( "P2" ) ) // ASCII PGM
     {
       int tmp;
-      while ( ( *in >> tmp ) && not( in->eof() ) )
+      while ( ( *in >> tmp ) and not( in->eof() ) )
```
Comment (jougs) [inline sli/sligraphics.cc:196]:
```suggestion
      while ( *in >> tmp and not in->eof() )
```

**Review by jougs — COMMENTED** [review]
I guess the replacement of the operators is unproblematic here. I still have a bunch of other comments regarding the statements in the lines that change. Most of my comments also apply to other places in the code in addition to the ones I have marked.

File: libnestutil/lockptr.h (hunk 1 above)
Comment (JanVogelsang) [reply libnestutil/lockptr.h:117]: Will be addressed in #2499

File: models/ac_generator.cpp (hunk 3 above)
Comment (JanVogelsang) [reply models/ac_generator.cpp:205]: Can create a PR where I replace those. I would rather do it once for all than just for some "random" occurrences.

File: sli/tarrayobj.cc (hunk 6 above)
Comment (JanVogelsang) [reply sli/tarrayobj.cc:633]: Done

File: models/ac_generator.cpp (hunk 3 above)
Comment (JanVogelsang) [reply models/ac_generator.cpp:205]: See #2506

File: sli/sligraphics.cc (hunk 14 above)
Comment (jougs) [reply sli/sligraphics.cc:196]: This was probably overlooked

**Review by jougs — APPROVED** [review]
Thanks for taking care of this. It is really much appreciated :-)

**Pull request merged**